packages/doenetml-worker-rust/lib-doenetml-core/src/core/props/data_query/content_filter.rs

```diff
@@ -0,0 +1,131 @@
+    /// Match components that have a prop matching a profile;
+    /// The value of the matched prop is tested against `Cond` and only matches
+    /// if `Cond` is true.
+    HasPropMatchingProfileAndCondition(PropProfile, Cond<PropValue>),
+    /// Match content is a component.
+    IsComponent,
+    /// Match content is a string.
+    IsString,
+}
+
+/// Information needed to determine of `node` matches a given filter condition.
+#[derive(Debug, Clone)]
+pub struct FilterData<'a> {
+    pub node: GraphNode,
+    pub origin: GraphNode,
+    pub document_model: &'a DocumentModel,
+}
+
+impl ApplyTest<FilterData<'_>, GraphNode> for ContentFilter {
+    fn apply_test(&self, data: &FilterData) -> bool {
+        // Warning: Only use `get_prop_unchecked` for getting a prop in this function.
+        // This is so that we possibly catch errors a little earlier if a prop that is not
+        // set in the dependency graph is accessed.
+        let node = data.node;
+        let document_model = data.document_model;
+        let origin = data.origin;
+
+        match self {
+            ContentFilter::IsType(type_name) => match node {
+                GraphNode::Component(_) => {
+                    let component_type = document_model.get_component_type(node);
+                    component_type.eq(type_name)
+                }
+                _ => false,
+            },
+            ContentFilter::HasPropMatchingProfile(profile) => match node {
+                GraphNode::Component(component_idx) => {
+                    let prop =
+                        document_model.get_component_prop_by_profile(component_idx, &[*profile]);
+                    prop.is_some()
+                }
+                _ => {
+                    // `PropProfile::String` is the only profile that non-component nodes can match.
+                    // We special case it here.
+                    if matches!(profile, PropProfile::String)
+                        && matches!(node, GraphNode::String(_))
+                    {
+                        return true;
+                    }
+                    false
+                }
+            },
+            ContentFilter::HasPropMatchingProfileAndCondition(profile, cond) => match node {
+                GraphNode::Component(component_idx) => {
+                    let prop =
+                        document_model.get_component_prop_by_profile(component_idx, &[*profile]);
+
+                    let prop = prop.map(|prop_pointer| {
+                        let prop_node = document_model.prop_pointer_to_prop_node(prop_pointer);
+                        document_model._get_prop_unchecked(prop_node, origin)
+                    });
+
+                    prop.map(|prop| {
+                        let value = prop.value;
+                        // Test the condition
+                        cond.apply_test(&value)
+                    })
+                    .unwrap_or(false)
+                }
+                _ => false,
```

Comment:
We need a `GraphNode::String(_)` match arm. If `profile` is `String` or `LiteralString`, then we need to get the value of the string as a `PropValue` and then apply `cond` to that value.